### ISSUE TITLE
moving hosting of artifacts out of jcenter (to maven central)

### ISSUE THREAD

**Issue by davidB** [jMonkeyEngine/jmonkeyengine#1468]
EOL of bintray/jcenter is planned for 2021/05/01 as explained in [Service End for Bintray, JCenter, GoCenter, and ChartCenter | JFrog](https://jfrog.com/blog/into-the-sunset-bintray-jcenter-gocenter-and-chartcenter/).

Proposed plan to move the release artifact out of bintray:

- [x] Prepare hosting on maven central (via OSS sonatype), (see [OSSRH Guide](https://central.sonatype.org/pages/ossrh-guide.html))
  - [x] create user account for release manager on https://issues.sonatype.org/
  - [x] request groupId `org.jmonkeyengine` project on oss.sonatype see [Producers](https://central.sonatype.org/pages/producers.html):
    - [x]  create a ticket on [OSSRH](https://issues.sonatype.org/secure/CreateIssue.jspa?issuetype=21&pid=10134)
      - summary : `hosting of org.jmonkeyengine`
      - description: `Request to host artifacts from jMonkeyEngine, currently hosted on bintray/jcenter (but not yet sync)`
      - groupId: `org.jmonkeyengine`
      - project Url: `https://jmonkeyengine.org/`
      - scm: `https://github.com/jMonkeyEngine/jmonkeyengine.git`
      - usernames: add `davidb` (to be able to sync from bintray) and user account previously created
      - already synced : None
    -  [x]  create a TXT record in your DNS referencing your OSSRH ticket number and your groupId will be approved on OSSRH
- [x] request maven central sync on bintray (via the old bintray UI)
   - eg: https://bintray.com/jmonkeyengine/org.jmonkeyengine/jme3-core/#central
   - if existing artifacts don't match the [requirements](http://central.sonatype.org/pages/requirements.html), sync will failed
- [x] update release workflow
   - [x] replace into scripts deploy release to bintray by deploy release to oss.sonatype (see deploy via [Gradle](https://central.sonatype.org/pages/gradle.html) )
   - [x] update doc for release manager
- [x] update documentation about repository where are hosted artifacts (README, wiki,...)
- [x] update dependencies of artifact that depend on jmonkeyengine

**Comment by davidB** [comment]
I can take care of some of the steps (on bintray and OSSRH, but not on DNS).

**Comment by stephengold** [comment]
Thank you for opening this issue and outlining the plan. FYI there is also a discussion topic at the JMonkeyEngine forum/hub: 
  https://hub.jmonkeyengine.org/t/sunset-on-may-1st-bintray-jcenter/44251

**Comment by stephengold** [comment]
What does "readme doc on" entail?

**Comment by riccardobl** [comment]
Thank you for opening the issue. 
I've also started doing some migrations as reported here: https://hub.jmonkeyengine.org/t/sunset-on-may-1st-bintray-jcenter/44251/36?u=riccardoblb
Regarding the DNS changes, i can do them, i just need to know what to do in case someone else is taking care of the process.

**Comment by davidB** [comment]
I created the request for `org.jmonkeyengine` on OSSRH: [[OSSRH-64318] hosting of org.jmonkeyengine - Sonatype JIRA](https://issues.sonatype.org/browse/OSSRH-64318) with @stephengold and @davidB as release manager.
Adding other "release manager" and gpg key could be done later (I did it for one of my hosted artifact), iirc they prefere to register key for individual than a generic key (like suggested on the forum). It will also be possible to remove my account as release manager.

@riccardobl can add a TXT record on DNS for org.jmonkeyengine with valkue `OSSRH-64318` ?

**Comment by MeFisto94** [comment]
One question: Do we get troubles as our packages used to be `com.jme3.x`?
I mean the artifact coords are `org.jmonkeyengine`, but the whole code actually isn't.
And we have no control over jme3.com, do we?

**Comment by davidB** [comment]
> One question: Do we get troubles as our packages used to be `com.jme3.x`?
> I mean the artifact coords are `org.jmonkeyengine`, but the whole code actually isn't.
> And we have no control over jme3.com, do we?

No troubles, I guess. I host artifact under groupId `net.alchim31.maven`, but the java packages are not under net.alchim31 see https://github.com/davidB/scala-maven-plugin/tree/master/src/main/java


> What does "readme doc on" entail?

I mean to udpate the README,... if they have reference to the repository from where to download artifacts

**Comment by riccardobl** [comment]
> @riccardobl can you add a TXT record on DNS for org.jmonkeyengine with value `OSSRH-64318` ?

For jmonkeyengine.org or for org.jmonkeyengine.jmonkeyengine.org ? I added this one:

![Screenshot from 2021-02-09 15:25:19](https://user-images.githubusercontent.com/4943530/107377317-1d844a80-6aeb-11eb-9ac0-7b21e87e1bb2.png)

**Comment by pspeed42** [comment]
Maven central doesn't care about package names.  When you reserver a groupId, they need some assurance that someone who actually owns the domain isn't going to come along tomorrow and claim it.

No pom.xml with com.jme3 as a groupId is a real release.  If these exist then we need to let them die.

**Comment by davidB** [comment]
jme3-core `3.3.2-stable` is sync from bintray to maven central (see https://repo1.maven.org/maven2/org/jmonkeyengine/jme3-core/)
Every artifacts and version will require manual sync. :-( I can do it for version 3.3.2-stable if you like.
@stephengold can you try to sync some, use your ossrh cred as token (key == username, password == password) ?

**Comment by pspeed42** [comment]
@davidB super cool.  I'm wondering if there are any tips you can provide the community for anyone also trying to move their libraries over?  Any pitfalls that we should watch out for, etc.?  (I have quite a few libraries to move myself and so have a vested interest.)

**Comment by stephengold** [comment]
@davidB I'll try doing some manual syncs and report back here.

**Comment by stephengold** [comment]
I tried syncing jme3-android v3.3.2-stable from Bintray, using my OSSRH username ("stephengold") and password, and I got "Authentication failed". I also tried syncing jme3-core v3.3.0-beta2 in the same manner and got the same result.

**Comment by davidB** [comment]
@pspeed42 yes I could provide a small instructions list to move, when we have a working flow.

@stephengold have you already try to connect/sign-in to https://oss.sonatype.org/ ? if you connect and retry to sync, if failed to authenticate, you can try with the access token provide at https://oss.sonatype.org/#profile;User%20Token (click on you name on top right > profile > User token (in the combobox) )

**Comment by stephengold** [comment]
I'm signed into Sonatype. Thanks for explaining how to generate a user token. I tried using that token to sync jme3-android v3.3.2-stable. This time I got a 403 error: "User 'stephengold' missing 'stage' permission for staging profile: 13769b099076e"

**Comment by davidB** [comment]
Do you have "org.jmonkeyengine" into the list of staging profile: https://oss.sonatype.org/#stagingProfiles ? (I requested to give you access in the ticket)

**Comment by stephengold** [comment]
No. I see only one staging profile, and that's for "com.github.stephengold". I tried refreshing; no change. What's the Sonatype ID of the ticket?

**Comment by davidB** [comment]
> No. I see only one staging profile, and that's for "com.github.stephengold". I tried refreshing; no change. What's the Sonatype ID of the ticket?

[OSSRH-64318] hosting of org.jmonkeyengine - Sonatype JIRA 

In this case, I'll request to add you and maybe @riccardobl and @pspeed42 after creation of account on sonatype.

**Comment by pspeed42** [comment]
I have an account on sonatype already from before: pspeed42

**Comment by davidB** [comment]
request to allow @pspeed42 and @stephengold sent: [[OSSRH-64383] add deployer for org.jmonkeyengine - Sonatype JIRA](https://issues.sonatype.org/browse/OSSRH-64383)

I allow made a script to sync artifacts, it may could help some of you (to sync your own projects or to sync other versions)
see [a script to batch request sync from bintray/JCenter into maven central](https://gist.github.com/davidB/0fe652acf9f6e560273541dd52774dbc)

I got the following issues:
- jme3-android required a special flow (semi automatic, with `"close": "0"` and manual release on oss.sonatype.org)
- jme3-testdata, jme3-ios, jme3-vr are not included into jcenter, I sent a request (pending)
- jme3-examples doesn't include javadoc => blocker for maven central

else version `3.3.2-stable` seems available on maven central (see https://repo1.maven.org/maven2/org/jmonkeyengine/), but I do not test it or try to build a project with artifact from maven central. Can one of you test it ?

**Comment by davidB** [comment]
maven central doesn't the check correctly dependencies, because it accept jme3-jbullet but dependencies are missing on maven central (iirc it was one of the cause we select bintray, some year ago)

**Comment by pspeed42** [comment]
We selected bintray because is was 100x easier.

**Comment by stephengold** [comment]
> maven central doesn't the check correctly dependencies

I've already uploaded several artifacts to MavenCentral with dependencies that weren't (yet) on MavenCentral. There doesn't seem to be a firm requirement in that regard.

**Comment by stephengold** [comment]
"org.jmonkeyengine" appear in my staging profiles now. I'll try copying the 3.2.4-stable release to MavenCentral.

**Comment by davidB** [comment]
Great permission was also granted for @pspeed42 .

**Comment by stephengold** [comment]
+ I copied 3.2.4-stable by hand, except for jme3-examples, jme3-ios, and jme3-testdata.
+ jme3-vr is linked to JCenter now, so I copied that for both 3.2.4-stable and 3.3.2-stable.

**Comment by stephengold** [comment]
Tonight I'm going to try using the script to copy 3.1.0-stable and 3.3.0-stable from JCenter to MavenCentral.

**Comment by stephengold** [comment]
Apparently jme3-vr didn't exist in JME 3.1.0. Other than that, the automation worked flawlessly, even for jme3-android. I didn't have to do anything manually.

What happened regarding linking jme3-ios to JCenter?

I'll try to copy the remaining "stable" releases tomorrow (3.2.0, 3.2.1, 3.2.2, and 3.2.3).

**Comment by stephengold** [comment]
jme3-android/3.2.0-stable lacks JavaDoc, as does jme3-core/3.2.0-stable and jme3-core/3.2.1-stable

3.2.2-stable and 3.2.3-stable seem to be in better shape.

**Comment by stephengold** [comment]
This is https://github.com/jMonkeyEngine/jmonkeyengine/issues/861 returning to bite us, I think.

**Comment by stephengold** [comment]
The automation script didn't work smoothly for 3.2.2 and 3.2.3, but progress was made. I'm going to retry some of the artifacts that didn't get copied, then post a status report to the Forum.
Since jme3-ios still isn't linked to JCenter, I've added a polite request of my own.
There are doubtless more artifacts that could be copied---alpha and beta releases, for instance---but I don't think they're very important.
It's time to start copying artifacts that aren't on JCenter (including Jaime and the libraries without javadoc) to the Minio instance.

**Comment by tlf30** [comment]
Will snapshots be available anywhere after jcenter goes down?
(Sometimes I feel like I am the only one who uses the jme snapshots)

**Comment by stephengold** [comment]
What snapshots are you referring to?

**Comment by tlf30** [comment]
I was referring to these snapshots, I guess I thought that they were on jcenter, but I see it is actually jitpack so I don't think they will be impacted: https://jitpack.io/#jMonkeyEngine/jmonkeyengine/-SNAPSHOT.

**Comment by stephengold** [comment]
I believe all the JCenter/Bintray artifacts that can be synced to MavenCentral without rework have been synced. That's everything in the "org.jmonkeyengine" repo except jme3-examples, jme3-testdata, and the older 3.2.x releases.

I've confirmed that https://artifacts.jmonkeyengine.org works as a Maven repository, but aside from one test build, nothing's been uploaded there yet.

Riccardo very kindly set up the Minio instance so we could use it to store native snapshots. However, I think we may be able to avoid using it for that purpose. I recently submitted [PR 1474](https://github.com/jMonkeyEngine/jmonkeyengine/pull/1474) to remove native Bullet from the Engine. Once that's integrated, jme3-android will be the only native build in JMonkeyEngine. Then, in order to address issue #1239, we might to go back to using pre-built natives in jme3-android, as we did prior to v3.3. At that point, there would be *no* native builds in the Engine. @riccardobl do you concur?

**Comment by stephengold** [comment]
It appears we'll still have native builds in the Engine even after #1239 is resolved, so I've begun drafting a commit to use objects.jmonkeyengine.org instead of dl.bintray.com for native snapshots.

I'm not very familiar with the Engine's build/CI process and feel I'm working out of my depth. Is anyone available to review my work and/or help if I get stuck?

**Comment by davidB** [comment]
I didn't touch the build/CI of this project since ~ 6y, so I should not be helpful on this point, but I could take a look if you need :eyes: of a :duck: .

**Comment by stephengold** [comment]
I've added an Actions secret to the GitHub repo. Its name is "OBJECTS_KEY", and it holds the secret key for the MinIO instance. This should provide our CI scripts at GitHub Actions with write access to that instance.

I think the next step will be upload native snapshots in parallel to both BinTray and the MinIO instance and verify that the results match.

**Comment by stephengold** [comment]
That seemed to work.
I think the next step will be to download natives snapshots from the MinIO instance.

**Comment by stephengold** [comment]
That seemed to work. I also removed uploading natives snapshots to Bintray, since that relies on creating new packages at Bintray, which might fail after 31 March.

I think the next step will be to deploy releases to SonaType OSSRH. This looks tricky because deployment to Bintray currently uses curl commands in custom shell functions. JME doesn't use Gradle's maven-publish plugin at all!

**Comment by stephengold** [comment]
I've added an Actions secret to the GitHub repo. Its name is "OSSRH_PASSWORD", and it holds the password for OSSRH deployment.

**Comment by stephengold** [comment]
JME deployment to Bintray produces PGP signatures (.asc files) for all the Maven artifacts, using RSA key 379CE192D401AB61. But JME doesn't use Gradle's "signing" plugin, and none of the Bash scripts invoke "gpg" or "gpg2". So where do the signatures come from?

It turns out that the "org.jmonkeyengine" repo at Bintray is configured to automatically sign uploaded files using Bintray's public/private key pair. Since I don't know Bintray's private key, I'll add my personal keys as Actions secrets and use those for signing.

**Comment by stephengold** [comment]
The build is broken tonight due to persistent "Bad gateway" errors (HTTP 502) jcenter.bintray.com
I'm unsure whether this is related to the impending sunset or not.
Developers who have spotbugs-gradle-plugin, groovy-all, and so on in their Gradle cache won't notice the issue right away.

**Comment by stephengold** [comment]
JMonkeyEngine v3.4.0-alpha3 was automatically deployed to both JCenter (via Bintray) and MavenCentral (via Sonatype).

Given Bintray's current unreliability, I think the next step should be to stop deploying releases to Bintray.

**Comment by stephengold** [comment]
The last 3 test releases (3.4.0-alpha6, -alpha7, and -beta1) all deployed successfully.

As far as I know, there's no existing documentation for the JME release process. To remedy this, I've opened a pull request: https://github.com/jMonkeyEngine/wiki/pull/128

**Comment by stephengold** [comment]
Still need to update https://jmonkeyengine.org/start/

**Comment by stephengold** [comment]
Submitted a pull request to fix the quickstart: https://github.com/jMonkeyEngine/jmonkeyengine-website/pull/17

**Comment by stephengold** [comment]
Submitted a pull request to fix part of the wiki: https://github.com/jMonkeyEngine/wiki/pull/131

**Comment by stephengold** [comment]
The old native snapshots at Bintray are no longer available, making it difficult to build the v3.3 branch. Here is a typical build failure:
```text
Executing: gradle :assemble
Arguments: [-c, /home/sgold/Git/33/jmonkeyengine/settings.gradle]


> Configure project :
Latest tag:v3.3.2-stable
Auto-detecting version
Revision: 7049
Hash: 1ab3e92d041ccbd3e6794f2c3c154febee43fb8e
Short Hash: 1ab3e92
Tag: null
Build Date: 2021-05-07
Build Branch: v3.3
Use commit hash as version false
Base Version: 3.3.2
Build Suffix: SNAPSHOT
Build Version: 3.3.2-SNAPSHOT
Use natives snapshot: https://dl.bintray.com/jmonkeyengine/files/9f026b41ca0cab34042262a830a3979107721da1/jme3-natives.zip

> Task :getNativesZipFile FAILED
Download natives from https://dl.bintray.com/jmonkeyengine/files/9f026b41ca0cab34042262a830a3979107721da1/jme3-natives.zip to /home/sgold/Git/33/jmonkeyengine/build/9f026b41ca0cab34042262a830a3979107721da1-natives.zip

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':getNativesZipFile'.
> Can't get https://dl.bintray.com/jmonkeyengine/files/9f026b41ca0cab34042262a830a3979107721da1/jme3-natives.zip to /home/sgold/Git/33/jmonkeyengine/build/9f026b41ca0cab34042262a830a3979107721da1-natives.zip
```

Fixing this would take a lot of work, so I hope we never need the v3.3 branch again.

**Comment by stephengold** [comment]
I've removed the last references to JCenter from the "master" and "v3.3" branches of the JME Wiki.

I still have concerns about the SDK, the jMonkeyEngine Contributions repos (including Lemur and zay-es), the Software Store, MonkeyZone, and the jMonkeyEngine-archive repos. However, I believe those projects have their own issue trackers. So this issue can be closed.
